src/main/scala/chisel3/util/BitPat.scala

```diff
@@ -224,8 +227,95 @@ package experimental {
         case _ => false
       }
     }
+
+    /**
+     * Calculate the inverse of this pattern set.
+     * 
+     * @return A BitSet matching all value (of the given with) iff it doesn't match this pattern.
+     */
+    def inverse: BitSet = {
+      val total = BitPat("b" + ("?".repeat(this.getWidth)))
+      total.subtract(this)
+    }
+
+    /**
+      * Calculate lowest posible value to match this pattern seset.
+      *
+      * @throws UnsupportedOperationException if this this is an empty pattern set.
+      * @return The matched value
+      */
+    def lowerbound: BigInt = terms.map(_.lowerbound).min
+
+    /**
+      * Convert arbitrary BitSet into continous ranges
+      * 
+      * @note Due to the potential large number of generated ranges, this method return an iterator.
+      * API user should be careful when the number of the ranges are not guarenteed.
+      *
+      * @return An iterator yielding all continous ranges in this pattern. The ranges are guaranteed to be increasing,
+      *     non-overlaping, and all of them will precisely cover the entire BitSet.
+      */
+    def toRanges: Iterator[BitSetRange] = new BitSetRangeIterator(this)
+  }
+
+  sealed class BitSetRange(val start: BigInt, val length: BigInt, val width: Int) extends BitSet { outer =>
+    override def terms = {
+      val collected = mutable.Set[BitPat]();
+      var ptr = start;
+      var left = length;
+      while (left > 0) {
+        var curPow = left.bitLength - 1
+        if (ptr != 0) {
+          val maxPow = ptr.lowestSetBit
+          if (maxPow < curPow) curPow = maxPow
+        }
+
+        val inc = BigInt(1) << curPow
+        require((ptr & inc - 1) == 0, "BitPatRange: Internal sanity check")
+        val mask = (BigInt(1) << width) - inc
+        collected.add(new BitPat(ptr, mask, width))
+        ptr += inc
+        left -= inc
+      }
+
+      collected.toSet
+    }
+
+    // TODO: width!
+    override def toString: String = s"BitSetRange(0x${start.toString(16)} - 0x${(start + length).toString(16)})"
+  }
+
+  object BitSetRange {
+    def apply(
+      start:          BigInt,
+      length:         BigInt,
+      width:          Int = -1,
+      allowUnaligned: Boolean = false
+    ): BitSetRange = {
+      require(length > 0, "Cannot construct a empty BitSetRange")
+      val max_known_length = (start + length - 1).bitLength
+      val w = if (width >= 0) width else max_known_length
+      require(w >= max_known_length, "Cannot construct a BitSetRange with width smaller than its range end")
+      val ret = new BitSetRange(start, length, w)
+
+      if (!allowUnaligned) require(ret.terms.size == 1, "Unaligned BitSetRange")
+
+      ret
+    }
```

Comment:
This factory is useful, but I think for simplicity, we can move this API to `BitSet.range(start: BitInt, length: BigInt)`.
`width` can be refactored as `BitSet.expandWidthTo(width: Int)`.
`allowUnaligned` can also be refactored as `BitSetRange.isAligned: Bool` function.

We may need(or even change to) another API: `BitSet.range(range: Range)`

Reply:
So how are we going to implement expandWidthTo? Fill in `?` or `0`? In the ranges case, we should obvious fill '0', but for more general bit patterns, this is not exactly clear what we should do.

IMHO two bit pattern of different width is two completely different thing with different schematic. There is no *obvious correct* way of converting a shorter one into a longer one.

Reply:
yeah, I agree, expanding width is tricky here. So maybe we still should keep the width as an arg.